### ISSUE TITLE
How does luarocks check wether a dependency exists ?

### ISSUE THREAD

**Issue by teto** [luarocks/luarocks#1659]
### Please fill out the following:
- **Platform:**
  - nixos

- **LuaRocks version:**
  - `luarocks --version` 3.11.0

- **Configuration file:**
```
external_deps_dirs = {
  "/nix/store/zyxxvyi22arcqj9a9pblncwany03308w-wrap-lua-hook",
  "/nix/store/3f964jpwhr5wpzy6lj5c2s2l9j799dbf-lua-5.2.4",
  "/nix/store/s83arnckylw48c29pvhq0z5g1hb62xnd-luarocks-check-hook",
  "/nix/store/4dkx3kb50z96v3cac5xnwcg7f1z4acxz-neovim-unwrapped-0.9.5"
}
local_cache = ""
rocksSubdir = "sqlite-v1.2.2-0-rocks"
rocks_subdir = "sqlite-v1.2.2-0-rocks"
rocks_trees = {
  {
    ["name"] = "current",
    ["rocks_dir"] = "current",
    ["root"] = "/nix/store/9yl7qjh8dpj5q2500ic2n993zj6p7d8c-sqlite-luarocks-config.lua"
  },
  {
    ["name"] = "luarocks-0",
    ["rocks_dir"] = "luarocks-3.11.0-rocks",
    ["root"] = "/nix/store/q49lxj6n5pid64ny5y95blxfvmn8563k-luarocks-3.11.0"
  },
  {
    ["name"] = "luv-1",
    ## note that this folder does not exist
    ["rocks_dir"] = "/nix/store/p40hd6l5ni9c90sv45f8kv4p47j7gizm-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks",
    # this one does, content afterwards:
    ["root"] = "/nix/store/p40hd6l5ni9c90sv45f8kv4p47j7gizm-lua5.2-luv-1.44.2-1"
  }
}
```

- **LuaRocks output from when the issue occurred:**
```
$ luarocks --verbose make --deps-mode=all --tree=$out sqlite....rockspec
os.execute:     mkdir -p '/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0'
..........: 0
fs.execute("ln", "/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/.lock.tmp.92634548", "/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/lockfile.lfs")
fs.quote_args("ln", "/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/.lock.tmp.92634548", "/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/lockfile.lfs")
fs.Q("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/.lock.tmp.92634548")
fs.Q("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/lockfile.lfs")
fs.execute_string("ln '/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/.lock.tmp.92634548' '/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/lockfile.lfs'")

os.execute:     ln '/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/.lock.tmp.92634548' '/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/lockfile.lfs'
..........: 0
fs.exists("/build/sqlite-v1.2.2-0.rockspec")
fs.absolute_name("/build/sqlite-v1.2.2-0.rockspec")
fs.absolute_name("/build/sqlite-v1.2.2-0.rockspec")
fs.absolute_name("/build/sqlite-v1.2.2-0.rockspec")
fs.exists("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2.zip")
fs.absolute_name("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2.zip")
fs.exists("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2")
fs.absolute_name("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2")
fs.exists("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest")
fs.absolute_name("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest")
fs.exists("current/manifest-5.2.zip")
fs.absolute_name("current/manifest-5.2.zip")
fs.exists("current/manifest-5.2")
fs.absolute_name("current/manifest-5.2")
fs.exists("current/manifest")
fs.absolute_name("current/manifest")
fs.exists("luarocks-3.11.0-rocks/manifest-5.2.zip")
fs.absolute_name("luarocks-3.11.0-rocks/manifest-5.2.zip")
fs.exists("luarocks-3.11.0-rocks/manifest-5.2")
fs.absolute_name("luarocks-3.11.0-rocks/manifest-5.2")
fs.exists("luarocks-3.11.0-rocks/manifest")
fs.absolute_name("luarocks-3.11.0-rocks/manifest")
fs.exists("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2.zip")
fs.absolute_name("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2.zip")
fs.exists("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2")
fs.absolute_name("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2")
fs.exists("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest")
fs.absolute_name("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest")
Missing dependencies for sqlite v1.2.2-0:
fs.exists("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2.zip")
fs.absolute_name("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2.zip")
fs.exists("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2")
fs.absolute_name("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest-5.2")
fs.exists("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest")
fs.absolute_name("/nix/store/8m025rihwm6cjcbrjf3mm7b97zmrc9s2-lua5.2-sqlite-v1.2.2-0/sqlite-v1.2.2-0-rocks/manifest")
fs.exists("current/manifest-5.2.zip")
fs.absolute_name("current/manifest-5.2.zip")
fs.exists("current/manifest-5.2")
fs.absolute_name("current/manifest-5.2")
fs.exists("current/manifest")
fs.absolute_name("current/manifest")
fs.exists("luarocks-3.11.0-rocks/manifest-5.2.zip")
fs.absolute_name("luarocks-3.11.0-rocks/manifest-5.2.zip")
fs.exists("luarocks-3.11.0-rocks/manifest-5.2")
fs.absolute_name("luarocks-3.11.0-rocks/manifest-5.2")
fs.exists("luarocks-3.11.0-rocks/manifest")
fs.absolute_name("luarocks-3.11.0-rocks/manifest")
fs.exists("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2.zip")
fs.absolute_name("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2.zip")
fs.exists("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2")
fs.absolute_name("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest-5.2")
fs.exists("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest")
fs.absolute_name("/nix/store/27h535935rlvq09a36iz1s59gjhbgxmp-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks/manifest")
   luv (not installed)

```

- **Description:**
In nixpkgs, the build of the latest sqlite fails because it cant find the dependency `luv`.
It is installed in a rocktree specified in the config though:
```
tree /nix/store/z4gg8r54niqvwih19zksnfsa68zxpsw5-lua5.2-luv-1.44.2-1/
$ /nix/store/z4gg8r54niqvwih19zksnfsa68zxpsw5-lua5.2-luv-1.44.2-1/
├── lib
│   ├── lua
│   │   └── 5.2
│   │       └── luv.so                   # < ONLY useful file 
│   └── luarocks
│       └── rocks-5.2
│           ├── luv
│           │   └── 1.44.2-1
│           │       ├── doc
│           │       │   ├── docs.md
│           │       │   ├── LICENSE.txt
│           │       │   └── README.md
│           │       ├── luv-1.44.2-1.rockspec
│           │       └── rock_manifest
│           └── manifest
└── nix-support
    └── propagated-build-inputs
```
so I wonder how does luarocks find dependencies ? I could patch luarocks to help myself, I find the logs a bit hard to interpret with just shell commands and without human-readable logs of the steps we are at.

**Comment by hishamhm** [comment]
> so I wonder how does luarocks find dependencies ? 

Reading from the manifest file (in your case, `lib/luarocks/rocks-5.2/manifest`). That serves as a cache of the overall tree structure. 

> ```  
>  ## note that this folder does not exist
> ["rocks_dir"] = "/nix/store/p40hd6l5ni9c90sv45f8kv4p47j7gizm-lua5.2-luv-1.44.2-1/luv-1.44.2-1-rocks",
> ```

Since that doesn't exist, perhaps you shouldn't be setting this value in your config then? `/nix/store/z4gg8r54niqvwih19zksnfsa68zxpsw5-lua5.2-luv-1.44.2-1/lib/luarocks/rocks-5.2` seems to be the correct value there.

That would have been the value derived by default, if you weren't setting `rocks_subdir = "sqlite-v1.2.2-0-rocks"` at the top. 

Since your "root" directories are already split by package (`/nix/store/...-lua-5.x-package-version/`), is it really necessary to change the `rocks_subdir`? It seems it's only adding another `...-package-version/package-version-rocks` level to your structure, and complicating your configuration, unless I'm missing something.

**Comment by teto** [comment]
thank you so much for the speedy answer. So there must always be a manifest (I saw today while investigating this one could run `luarocks make --no-manifest` which I was tempted to enable) ?


> Since your "root" directories are already split by package (/nix/store/...-lua-5.x-package-version/), is it really necessary to change the rocks_subdir? It seems it's only adding another ...-package-version/package-version-rocks level to your structure, and complicating your configuration, unless I'm missing something.

In some cases, we (nixpkgs) create the union-of/merge those paths /nix/store/package1 /nix/store/package2 into a /nix/store/lua-env-package1-and-2 that is a forest of symlinks towards those packages (https://nixos.org/manual/nixpkgs/stable/#trivial-builder-symlinkJoin). The comments in code says we specify `rocksSubdir` to avoid collision between these files when creating this env.

As per your comment, seems like we generate the rocks_dir wrong. I will look into it. What's weird is that it seems to cause an issue just for that package so I wonder if luarocks doesn't have fallbacks for finding packages other than the manifest.

**Comment by hishamhm** [comment]
> So there must always be a manifest 

Yes, if you want it to perform dependency resolution.

>  I wonder if luarocks doesn't have fallbacks for finding packages other than the manifest.

It does not.

If you build all your packages and your build system tracks dependency management on its own (which seems to be the case), you could always use `--no-deps` and have Nix fetch the dependent packages itself, effectively using LuaRocks only as a build tool.

**Comment by alerque** [comment]
@teto Just as a quick point of reference, Arch Linux builds packages from LuaRocks by using `luarocks make --no-manifest` and setting the `--tree` to the empty destination location where files are then collected that belong to the package. Then we have a post-install hook for any packages that place/change files in the Lua lib directories that generates a new system wide manifest after any relevant install/upgrade/uninstall.

**Comment by teto** [comment]
@alerque much appreciated. I'll try to achieve sthg similar

**Comment by teto** [comment]
@alerque would you have a link towards the arch code that does what you suggest ? especially the manifest generation please

**Comment by alerque** [comment]
@teto of course. [Here are the `luarocks` package sources](https://gitlab.archlinux.org/archlinux/packaging/packages/luarocks). The `*.hook` files get placed in a location that the package manager automatically uses them. The rest is pretty straight forward. One per Lua version supported obviously because there might be different packages installed in each tree.

After any package install, upgrade, or removal, look fel files in this path:

```
Target = usr/lib/luarocks/rocks-5.4/*/**
```

If any files in that location got touched, generate a new manifest:

```
Exec = /usr/bin/luarocks-admin make_manifest --local-tree --lua-version=5.4 /usr/lib/luarocks/rocks-5.4/
```

**Comment by teto** [comment]
hhaaa I see it's using luarocks-admin nice, I was not sure if I had to write the code to generate the manifest. Looks pretty doable ty.

**Comment by teto** [comment]
Because the nix luarocks package doesn't have a manifest, marc had the following issue https://github.com/NixOS/nixpkgs/issues/316009 so I tried to generate the manifest for an installed `luarocks` with `luarocks-admin make-manifest`. I had to browse the source (I wish `luarocks` would print more about what it does, I always end up adding "print" because --verbose just prints cp/mv/rm without telling us why, is this a choice not to have loglevels like `luarocks --log-level=info` ?) to understand what's going on.

Correct me if I am wrong:
1. So make-manifest expects a `repository` which means a toplevel package with lua packages underneath
2. looking at the children folder, it tries to find rockspec or rocks and generate the manifest from there

In my installed version of luarocks, I dont have the rockspec available in the folder so make-manifest is unusable ?

NB: https://github.com/luarocks/luarocks/issues/1659#issuecomment-2045371038 calls make-manifest with `--lua-version` but it doesn't appear in --help. The default behavior tries to generate a manifest for every version (except with --local-tree), which generated warnings in my case (couldn't remove old unexiting manifests I think)
